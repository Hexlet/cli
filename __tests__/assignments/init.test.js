// @ts-check

const os = require('os');
const path = require('path');
const fsp = require('fs/promises');
const fse = require('fs-extra');
const nock = require('nock');
const git = require('isomorphic-git');

const assignmentsInitCmd = require('../../src/commands/assignments-init.js');
const { getValidator } = require('../../src/validator/index.js');
const { getFixturePath, readDirP } = require('../helpers/index.js');

const githubToken = 'some-github-token';
const hexletToken = 'some-hexlet-token';
const commandParts = ['assignments', 'init'];
const args = {
  githubToken, hexletToken, _: commandParts,
};
const githubUser = 'exampleuser';
const repoName = 'hexlet-assignments';
const secretName = 'HEXLET_TOKEN';
const branch = 'main';

const mockServerHost = 'localhost:8888';
const baseUrl = `http://${mockServerHost}`;

describe('program', () => {
  let customSettings;
  let hexletDir;

  beforeAll(() => {
    nock.disableNetConnect();
    nock.enableNetConnect(new RegExp(mockServerHost));

    const scope = nock('https://api.github.com').persist();

    scope
      .get('/')
      .reply(200, {}, {
        'x-oauth-scopes': 'repo, workflow',
      });
    scope
      .get('/user')
      .reply(200, { login: githubUser });
    scope
      .get(`/repos/${githubUser}/${repoName}`)
      .reply(404);
    scope
      .post('/user/repos')
      .reply(201, {
        html_url: `${baseUrl}/assignments-empty`,
        clone_url: `${baseUrl}/assignments-empty.git`,
      });
    scope
      .get(`/repos/${githubUser}/${repoName}/actions/secrets/public-key`)
      .reply(200, {
        key_id: '012345678912345678',
        key: 'pTkCe8klWSqGB9ZYJ+nLJknKMjfcfHWnU2irP+AS5zw=',
      });
    scope
      .put(`/repos/${githubUser}/${repoName}/actions/secrets/${secretName}`)
      .reply(201);
    scope
      .get(`/repos/${githubUser}/${repoName}/branches/${branch}`)
      .reply(404);

    git.push = jest.fn(() => {});
  });

  beforeEach(async () => {
    const tmpDir = await fsp.mkdtemp(path.join(os.tmpdir(), 'hexlet-cli-'));
    customSettings = { homedir: tmpDir };
    hexletDir = path.join(tmpDir, 'learning', 'Hexlet');
  });

  it('init without local config', async () => {
    const result = await assignmentsInitCmd.handler({ ...args, hexletDir }, customSettings);

    const actualConfig = await fse.readJson(result.hexletConfigPath);
    const expectedConfig = await fse.readJson(getFixturePath('assignmentsConfig.json'));
    expectedConfig.hexletDir = hexletDir;
    expect(actualConfig).toEqual(expectedConfig);

    const validate = getValidator('assignments');
    expect(validate(actualConfig)).toBeTruthy();

    expect(await readDirP(hexletDir)).toMatchSnapshot();
  });

  it('init with program config', async () => {
    await fse.copy(
      getFixturePath('programConfig.json'),
      path.join(customSettings.homedir, 'Hexlet', '.config.json'),
    );
    const result = await assignmentsInitCmd.handler({ ...args, hexletDir }, customSettings);

    const actualConfig = await fse.readJson(result.hexletConfigPath);
    const expectedConfig = await fse.readJson(getFixturePath('config.json'));
    expectedConfig.hexletDir = hexletDir;
    expect(actualConfig).toEqual(expectedConfig);

    const validate = getValidator('assignments');
    expect(validate(actualConfig)).toBeTruthy();

    expect(await readDirP(hexletDir)).toMatchSnapshot();
  });
});
