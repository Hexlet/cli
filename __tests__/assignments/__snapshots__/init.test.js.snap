// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignments init with existing repository & init again after changes 1`] = `
Array [
  Object {
    "name": ".current.json",
    "path": "hexlet-assignments/.current.json",
    "size": 39n,
  },
  Object {
    "name": "hexlet-check.yml",
    "path": "hexlet-assignments/.github/workflows/hexlet-check.yml",
    "size": 600n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/README.md",
    "size": 9233n,
  },
  Object {
    "name": ".gitignore",
    "path": "hexlet-assignments/basics-course/example/.gitignore",
    "size": 20n,
  },
  Object {
    "name": "Makefile",
    "path": "hexlet-assignments/basics-course/example/Makefile",
    "size": 23n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/basics-course/example/README.md",
    "size": 402n,
  },
  Object {
    "name": "file1",
    "path": "hexlet-assignments/basics-course/example/file1",
    "size": 12n,
  },
  Object {
    "name": "Makefile",
    "path": "hexlet-assignments/basics-course/start/Makefile",
    "size": 23n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/basics-course/start/README.md",
    "size": 402n,
  },
]
`;

exports[`assignments init with existing repository & init again after changes 2`] = `
Array [
  Object {
    "name": ".current.json",
    "path": "hexlet-assignments/.current.json",
    "size": 39n,
  },
  Object {
    "name": "hexlet-check.yml",
    "path": "hexlet-assignments/.github/workflows/hexlet-check.yml",
    "size": 600n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/README.md",
    "size": 9233n,
  },
  Object {
    "name": ".gitignore",
    "path": "hexlet-assignments/basics-course/example/.gitignore",
    "size": 20n,
  },
  Object {
    "name": "Makefile",
    "path": "hexlet-assignments/basics-course/example/Makefile",
    "size": 23n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/basics-course/example/README.md",
    "size": 402n,
  },
  Object {
    "name": "addedFile",
    "path": "hexlet-assignments/basics-course/example/addedFile",
    "size": 7n,
  },
  Object {
    "name": "file1",
    "path": "hexlet-assignments/basics-course/example/file1",
    "size": 12n,
  },
  Object {
    "name": "Makefile",
    "path": "hexlet-assignments/basics-course/start/Makefile",
    "size": 23n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/basics-course/start/README.md",
    "size": 402n,
  },
]
`;

exports[`assignments init with program config 1`] = `
Array [
  Object {
    "name": "hexlet-check.yml",
    "path": "hexlet-assignments/.github/workflows/hexlet-check.yml",
    "size": 600n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/README.md",
    "size": 9233n,
  },
]
`;

exports[`assignments init without local config 1`] = `
Array [
  Object {
    "name": "hexlet-check.yml",
    "path": "hexlet-assignments/.github/workflows/hexlet-check.yml",
    "size": 600n,
  },
  Object {
    "name": "README.md",
    "path": "hexlet-assignments/README.md",
    "size": 9233n,
  },
]
`;
